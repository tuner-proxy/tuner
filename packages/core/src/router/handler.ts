import type { TunerRequest } from '../shared/types';
import type { ConnectRequest } from '../wrapper/ConnectRequest';
import type { HTTPRequest } from '../wrapper/HTTPRequest';
import type { UpgradeRequest } from '../wrapper/UpgradeRequest';

export type Awaitable<T> = T | PromiseLike<T>;

export type RouteElement = string | RouteHandleElement | RouteElement[];

export type RouteHandleElement =
  | HTTPProcessFn
  | RouteHandler
  | null
  | false
  | undefined;

export type RouteNextFn = (
  handlers?: Array<HTTPProcessFn | RouteHandler>,
) => Promise<void>;

export type RequestProcessFn<T extends TunerRequest> = (
  req: T,
  next: RouteNextFn,
) => Awaitable<void>;

export type HTTPProcessFn = RequestProcessFn<HTTPRequest>;
export type ConnectProcessFn = RequestProcessFn<ConnectRequest>;
export type UpgradeProcessFn = RequestProcessFn<UpgradeRequest>;

export interface BaseRouteHandler {
  type: 'process-item';
}

export interface HTTPRouteHandler extends BaseRouteHandler {
  kind: 'common';
  process: HTTPProcessFn;
}

export interface ConnectRouteHandler extends BaseRouteHandler {
  kind: 'connect';
  process: ConnectProcessFn;
}

export interface UpgradeRouteHandler extends BaseRouteHandler {
  kind: 'upgrade';
  process: UpgradeProcessFn;
}

export type RouteHandler =
  | HTTPRouteHandler
  | ConnectRouteHandler
  | UpgradeRouteHandler;

export const httpHandler = getFactory<HTTPRouteHandler>('common');
export const connectHandler = getFactory<ConnectRouteHandler>('connect');
export const upgradeHandler = getFactory<UpgradeRouteHandler>('upgrade');

function getFactory<T extends RouteHandler>(kind: T['kind']) {
  return (process: T['process']) =>
    ({
      type: 'process-item',
      kind,
      process,
    }) as T;
}
