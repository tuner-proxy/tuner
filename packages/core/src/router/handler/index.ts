import { Duplex } from 'stream';

import { ConnectRequest } from '../../wrapper/ConnectRequest';
import { HTTPRequest } from '../../wrapper/HTTPRequest';
import { HTTPResponse, HTTPResponseOptions } from '../../wrapper/HTTPResponse';
import { UpgradeRequest } from '../../wrapper/UpgradeRequest';

export type RouteElement = string | RouteHandler | RouteElement[];

export type RouteHandler =
  | HTTPProcessFn
  | HTTPResponseOptions
  | ProcessItem
  | null
  | false
  | undefined;

export interface ProcessItem<TType = string, TProcess = ProcessFn> {
  type: TType;
  process: TProcess;
}

export type ProcessFn = (req: any, next: NextFn<any>) => MaybePromise<any>;

export interface NextOptions {
  final?: boolean;
}

export type NextFn<T> = (options?: NextOptions) => Promise<T>;

export type MaybePromise<T> = T | Promise<T>;

export type HTTPResult =
  | HTTPResponse
  | HTTPResponseOptions
  | undefined
  | null
  | void;

export type HTTPProcessItem = ProcessItem<'common', HTTPProcessFn>;

export type HTTPProcessFn = (
  req: HTTPRequest,
  next: NextFn<HTTPResponse>,
) => MaybePromise<
  | HTTPResult
  | HTTPProcessFn
  | HTTPProcessItem
  | Array<HTTPResult | HTTPProcessFn | HTTPProcessItem>
>;

export const httpHandler = getFactory<HTTPProcessItem>('common');

export type ConnectResult = Duplex | undefined | null | void;

export type ConnectProcessItem = ProcessItem<'connect', ConnectProcessFn>;

export type ConnectProcessFn = (
  req: ConnectRequest,
  next: NextFn<ConnectResult>,
) => MaybePromise<ConnectResult | ConnectProcessItem | ConnectProcessItem[]>;

export const connectHandler = getFactory<ConnectProcessItem>('connect');

export type UpgradeProcessItem = ProcessItem<'upgrade', UpgradeProcessFn>;

export type UpgradeProcessFn = (
  req: UpgradeRequest,
  next: NextFn<ConnectResult>,
) => MaybePromise<ConnectResult | UpgradeProcessItem | UpgradeProcessItem[]>;

export const upgradeHandler = getFactory<UpgradeProcessItem>('upgrade');

function getFactory<T extends ProcessItem>(type: string) {
  return (process: T['process']) => ({ type, process }) as any as T;
}
