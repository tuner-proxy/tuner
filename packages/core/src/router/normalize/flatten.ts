import flatMap from 'lodash.flatmap';

import type { RouteElement, RouteHandleElement } from '../handler';

export interface FlattenRoute {
  patterns: string[];
  handler: RouteHandleElement;
}

export function flattenRoutes(routes: RouteElement[]) {
  const result: FlattenRoute[] = [];
  for (let i = 0, ii = routes.length; i < ii; i += 1) {
    const fragments: string[] = [];
    for (; i < ii && typeof routes[i] === 'string'; i += 1) {
      fragments.push(routes[i] as string);
    }
    if (!fragments.length) {
      fragments.push('');
    }
    const activeFragments = fragments.filter((i) => i[0] !== '!');
    if (!activeFragments.length) {
      continue;
    }
    const handler = routes[i] as RouteHandleElement;
    let children: FlattenRoute[] = [{ patterns: [''], handler }];
    if (Array.isArray(handler)) {
      children = flattenRoutes(handler);
    }
    for (const child of children) {
      result.push({
        patterns: flatMap(child.patterns, (childPattern) =>
          activeFragments.map((frag) => frag + childPattern),
        ),
        handler: child.handler,
      });
    }
  }
  return result;
}
