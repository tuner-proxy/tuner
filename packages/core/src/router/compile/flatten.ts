import flatMap from 'lodash.flatmap';

import { RouteElement, RouteHandler } from '../handler';

export interface RouteItem {
  patterns: string[];
  handlers: RouteHandler[];
}

export function flattenRoutes(routes: RouteElement[]) {
  const result: RouteItem[] = [];
  for (let i = 0, ii = routes.length; i < ii; i += 1) {
    const fragments: string[] = [];
    for (; i < ii && typeof routes[i] === 'string'; i += 1) {
      fragments.push(routes[i] as string);
    }
    if (!fragments.length) {
      fragments.push('');
    }
    const activeFragments = fragments.filter((i) => i[0] !== '!');
    if (!activeFragments.length) {
      continue;
    }
    const handler = routes[i] as RouteHandler;
    let children: RouteItem[] = [{ patterns: [''], handlers: [handler] }];
    if (Array.isArray(handler)) {
      children = flattenRoutes(handler);
    }
    for (const child of children) {
      result.push({
        patterns: flatMap(child.patterns, (childPattern) =>
          activeFragments.map((frag) => frag + childPattern),
        ),
        handlers: child.handlers,
      });
    }
  }
  return result;
}
