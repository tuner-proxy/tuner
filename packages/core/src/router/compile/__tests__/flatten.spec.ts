import { RouteElement } from '../../handler';
import { flattenRoutes } from '../flatten';

describe('flatten', () => {
  test('common', ['//example.com', { status: 200 }]);

  test('multiple handlers', [
    '//example.com',
    [
      () => {
        // noop
      },
      { status: 404 },
    ],
  ]);

  test('with child routes', [
    '//example.com',
    '//demo.com',
    [{ status: 200 }, '/not-found', { status: 404 }],
  ]);

  test('disable fragment', ['//example.com', '!//demo.com', { status: 200 }]);

  test('disable fragment', ['//example.com', ['!', { status: 200 }]]);
});

function test(name: string, routes: RouteElement[]) {
  it(name, () => {
    expect(flattenRoutes(routes)).toMatchSnapshot();
  });
}
