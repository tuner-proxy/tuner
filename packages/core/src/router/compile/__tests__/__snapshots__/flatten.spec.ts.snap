// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flatten common 1`] = `
Array [
  Object {
    "handlers": Array [
      Object {
        "status": 200,
      },
    ],
    "patterns": Array [
      "//example.com",
    ],
  },
]
`;

exports[`flatten disable fragment 1`] = `
Array [
  Object {
    "handlers": Array [
      Object {
        "status": 200,
      },
    ],
    "patterns": Array [
      "//example.com",
    ],
  },
]
`;

exports[`flatten disable fragment 2`] = `Array []`;

exports[`flatten multiple handlers 1`] = `
Array [
  Object {
    "handlers": Array [
      [Function],
    ],
    "patterns": Array [
      "//example.com",
    ],
  },
  Object {
    "handlers": Array [
      Object {
        "status": 404,
      },
    ],
    "patterns": Array [
      "//example.com",
    ],
  },
]
`;

exports[`flatten with child routes 1`] = `
Array [
  Object {
    "handlers": Array [
      Object {
        "status": 200,
      },
    ],
    "patterns": Array [
      "//example.com",
      "//demo.com",
    ],
  },
  Object {
    "handlers": Array [
      Object {
        "status": 404,
      },
    ],
    "patterns": Array [
      "//example.com/not-found",
      "//demo.com/not-found",
    ],
  },
]
`;
